src/v3/bandwidthRtc.ts

```diff
@@ -346,29 +378,88 @@ export class BandwidthRtc {
     });
   }
 
-  private setupPublishingPeerConnection(): RTCPeerConnection {
+  private async addHeartbeatDataChannel(peerConnection: RTCPeerConnection): Promise<RTCDataChannel> {
+    //Create a heartbeat data channel
+    let heartbeatDataChannel = peerConnection.createDataChannel('__heartbeat__', {
+      id: 0,
+      negotiated: true,
+      protocol: 'udp'
+    });
+    heartbeatDataChannel.onmessage = function(event) {
+      logger.debug("Heartbeat Received:", event.data);
+      if (event.data === 'PING') {
+        heartbeatDataChannel.send("PONG");
+      }
+    }
+    return heartbeatDataChannel;
+  }
+
+  private async addDiagnosticDataChannel(peerConnection: RTCPeerConnection): Promise<RTCDataChannel> {
+    //Create a diagnostics data channel
+    let diagnosticsDataChannel = peerConnection.createDataChannel('__diagnostics__', {
+      id: 1,
+      negotiated: true,
+      protocol: 'udp'
+    });
+    diagnosticsDataChannel.onmessage = function(event) {
+      logger.info("Diagnostics Received:", event.data);
+    }
+    return diagnosticsDataChannel;
+  }
+
+  private async setupPublishingPeerConnection(): Promise<RTCPeerConnection> {
     logger.debug("Setting up publishing RTCPeerConnection");
     this.publishingPeerConnection = this.createPeerConnection();
 
     this.setupNewPeerConnection(this.publishingPeerConnection, this.setupPublishingPeerConnection);
+    // Attempt to restart ice if connection fails
+    this.publishingPeerConnection!.onconnectionstatechange = async (event: Event) => {
+      const pc = event.target as RTCPeerConnection;
+      let connectionState = pc.connectionState;
+      if (connectionState === "failed") {
+        await this.offerPublishSdp(true);
+        connectionState = pc.connectionState;
+        //TODO: add timeout so we dont loop here forever
+        while (connectionState === "failed") {
+          await new Promise( resolve => setTimeout(resolve, 5000) );
+          //Dont block on this, we should try multiple times
+          this.offerPublishSdp(true);
+          connectionState = pc.connectionState;
+        }
+      }
+    };
+
+    const heartbeatDataChannel = await this.addHeartbeatDataChannel(this.publishingPeerConnection);
+    this.publishedDataChannels.set(heartbeatDataChannel.label, heartbeatDataChannel);
+    this.publishHeartbeatDataChannel = heartbeatDataChannel;
+    const diagnosticDataChannel = await this.addDiagnosticDataChannel(this.publishingPeerConnection);
+    this.publishedDataChannels.set(diagnosticDataChannel.label, diagnosticDataChannel);
+    this.publishDiagnosticDataChannel = diagnosticDataChannel;
+    await this.offerPublishSdp();
```

Comment:
The first time `publish` is called `offerPublishSdp` is going to be called twice in short succession; seems like it'd be preferable to call that once, but I also see why we'd want to go ahead and get the ball rolling with the data channels here.